### PR TITLE
Introduce retries to the build script

### PR DESCRIPTION
**Pull request by j-maynard**
Sometimes the integration tests will fail when building SAML-Soap-proxy and SAML-proxy.  This seems to be due to a bottle next in available system resources when this first tries to build.  Attempting to run startup again most times seems to result in a successful build.  So these changes attempt to retry the build if it fails so the user doesn't have to.

This also includes the Fixes for moving everything to GHCR

### CODE REVIEW

File: build-local.rb [hunk 1]
```diff
@@ -19,29 +19,35 @@
 
 USAGE = <<ENDUSAGE
 Usage:
-    build-local -y yaml-file [-t count] [-h]
+    build-local -y yaml-file [-t count] [-r retires] [-h]
```
Comment (instantiator) [inline build-local.rb:22]:
```suggestion
    build-local -y yaml-file [-t count] [-r retries] [-h]
```

File: build-local.rb [hunk 2]
```diff
@@ -19,29 +19,35 @@
...
+                            not retry.
     -h, --help              Show's this help message
 ENDHELP
 
 # This is our worker thread which builds our docker images
 class ImageBuilder
   attr_accessor :image_var, :messages
 
-  def initialize(repo_name, config, spinner, images, script_dir)
+  def initialize(repo_name:, config:, spinner:, images:, script_dir:, retires:, write_success_log: false)
```
Comment (instantiator) [inline build-local.rb:41]:
```suggestion
  def initialize(repo_name:, config:, spinner:, images:, script_dir:, retries:, write_success_log: false)
```

File: build-local.rb [hunk 3]
```diff
@@ -19,29 +19,35 @@
...
     @repo_name = repo_name
     @config = config
     @spinner = spinner
     @image_var = ""
     @script_dir = script_dir
     @success = false
     @release = "verify-local-startup dev"
+    @retries = retires
```
Comment (instantiator) [inline build-local.rb:49]:
```suggestion
    @retries = retries
```

File: build-local.rb [hunk 4]
```diff
@@ -99,7 +112,7 @@ def success?
   end
 end
 
-def create_docker_images(thread_count, repos)
+def create_docker_images(thread_count, repos, retires)
```
Comment (instantiator) [inline build-local.rb:115]:
```suggestion
def create_docker_images(thread_count, repos, retries)
```

File: build-local.rb [hunk 5]
```diff
@@ -114,7 +127,14 @@ def create_docker_images(thread_count, repos)
...
     spinner = loading_spinners.register("[:spinner] #{repo_name}", format: :dots, success_mark: "#{thread_success_marks.sample}", error_mark: "😡")
     spinner.auto_spin
-    buildImage = ImageBuilder.new(repo_name, config, spinner, images, script_dir)
+    buildImage = ImageBuilder.new(
+        repo_name: repo_name,
+        config: config,
+        spinner: spinner,
+        images: images,
+        script_dir: script_dir,
+        retires: retires,
```
Comment (instantiator) [inline build-local.rb:136]:
```suggestion
        retries: retries,
```

File: build-local.rb [hunk 6]
```diff
@@ -163,6 +183,7 @@ def main()
       when '-h', '--help'         then args[:help] = true
       when '-y', '--yaml-file'    then next_arg = :yaml
       when '-t', '--threads'      then next_arg = :threads
+      when '-r', '--retry-build'  then next_arg = :retires
```
Comment (instantiator) [inline build-local.rb:186]:
```suggestion
      when '-r', '--retry-build'  then next_arg = :retries
```

File: build-local.rb [hunk 7]
```diff
@@ -189,6 +210,9 @@ def main()
   
   # Setup thread count
   thread_count = args[:threads].to_i
+  # Setup retries
+  retries = args[:retris].to_i
```
Comment (instantiator) [inline build-local.rb:214]:
```suggestion
  retries = args[:retries].to_i
```

**Review by instantiator — DISMISSED** [review]
Basically approved. Nothing but a small typo.

**Review by instantiator — DISMISSED** [review]

File: build-local.rb [hunk 8]
```diff
@@ -55,10 +63,21 @@ def build_image()
         2>&1"
     output = `#{cmd}`
     if $?.success?
-      @spinner.success
+      if @write_success_log
+        @spinner.success(" - see #{@script_dir}/logs/#{@repo_name}_build.log")
+        @putput = @output + output + "\nBuild failed... Unable to retry.\n"
```
Comment (jkunle) [inline build-local.rb:68]: Is this a typo `@putput` instead of `@output`

File: build-local.rb [hunk 9]
```diff
@@ -55,10 +63,21 @@ def build_image()
...
+        @spinner.success
+      end
       @image_var = "#{@config['image_env_var']}=#{image_name}\n"
       @success = true
+    elsif @retries > 0
+        @output = @output + output + "\nBuild failed... Retrying...\n"
+        @retries = @retries - 1
+        build_image()
     else
+      @putput = @output + output + "\nBuild failed... Unable to retry.\n"
```
Comment (jkunle) [inline build-local.rb:80]: Is this a typo @putput instead of @output

File: build-local.rb (hunk 8 above)
Comment (j-maynard) [reply build-local.rb:68]: Fixed

File: build-local.rb (hunk 9 above)
Comment (j-maynard) [reply build-local.rb]: Fixed

File: build-local.rb [hunk 10]
```diff
@@ -55,10 +63,21 @@ def build_image()
...
+        @spinner.success(" - see #{@script_dir}/logs/#{@repo_name}_build.log")
+        @output = @output + output + "\nBuild failed... Unable to retry.\n" 
+        File.write("#{@script_dir}/logs/#{@repo_name}_build.log", "log from command=#{cmd}\n#{output}", mode: "w")
+      else
+        @spinner.success
+      end
       @image_var = "#{@config['image_env_var']}=#{image_name}\n"
       @success = true
+    elsif @retries > 0
+        @output = @output + output + "\nBuild failed... Retrying...\n"
```
Comment (jkunle) [inline build-local.rb:76]: There is some extra indentation here
```@output = @output + output + "\nBuild failed... Retrying...\n" is equivalent to
@output += output + "\nBuild failed... Retrying...\n" ```
and 
```@retries = @retries - 1``` could be ```@retries -= 1```

File: build-local.rb [hunk 11]
```diff
@@ -55,10 +63,21 @@ def build_image()
...
+        @retries = @retries - 1
+        build_image()
```
Comment (jkunle) [inline build-local.rb:78]: Don't need the parenthesis

File: build-local.rb [hunk 12]
```diff
@@ -189,6 +217,11 @@ def main()
   
   # Setup thread count
   thread_count = args[:threads].to_i
+  # Setup retries
+  retries = args[:retris].to_i
```
Comment (jkunle) [inline build-local.rb:221]: retries = args[:retries].to_i

File: build-local.rb [hunk 13]
```diff
@@ -154,15 +180,17 @@ def generate_env(images)
...
+  args = { :yaml=>'repos.yml', :thread_count=>0, :retries=>1, :write_success_log=>false }
   unflagged_args = []
   next_arg = unflagged_args.first
 
   ARGV.each do |arg|
     case arg
-      when '-h', '--help'         then args[:help] = true
-      when '-y', '--yaml-file'    then next_arg = :yaml
-      when '-t', '--threads'      then next_arg = :threads
+      when '-h', '--help'               then args[:help] = true
```
Comment (jkunle) [inline build-local.rb:189]:
```
  if ! File.file?(args[:yaml])
    puts "Yaml file does not exist.  Exiting..."
    puts USAGE
    exit 1
  end
```
Use Unless
```
  unless File.file?(args[:yaml])
    puts "Yaml file does not exist.  Exiting..."
    puts USAGE
    exit 1
  end
```

File: build-local.rb [hunk 14]
```diff
@@ -189,6 +217,11 @@ def main()
   
   # Setup thread count
   thread_count = args[:threads].to_i
+  # Setup retries
+  retries = args[:retris].to_i
+
+  write_success_log = args[:write_success_log]
+
   if OS.mac? && thread_count == 0
```
Comment (jkunle) [inline build-local.rb:225]: `if OS.mac? && thread_count.zero?`

Comment (jkunle) [reply build-local.rb:225]: `threads.each { |t| t.join }` is equivalent to `threads.each(&:join)`

**Review by jkunle — APPROVED** [review]

**Review by instantiator — APPROVED** [review]

**Pull request merged**